examples/vision/siamesenetwork.py

```diff
@@ -0,0 +1,398 @@
+    Visualizes a few triplets from the supplied batches.
+    """
+
+    def show(ax, image):
+        ax.imshow(image)
+        ax.get_xaxis().set_visible(False)
+        ax.get_yaxis().set_visible(False)
+
+    fig = plt.figure(figsize=(9, 9))
+
+    axs = fig.subplots(3, 3)
+    for i in range(3):
+        show(axs[i, 0], anchor[i])
+        show(axs[i, 1], positive[i])
+        show(axs[i, 2], negative[i])
+
+
+"""
+# Load the dataset
+
+We are going to use the [Totally Looks Like dataset](https://drive.google.com/drive/folders/1qQJHA5m-vLMAkBfWEWgGW9n61gC_orHl). We are going to download it and unzip it inside the `~/.keras` directory.
+
+The dataset consists on two separate files:
+* `left.zip` contains the images that we will use as the anchor.
+* `right.zip` contains the images that we will use as the positive sample (an image that looks like the anchor.)
+"""
+
+cache_dir = Path(Path.home()) / ".keras"
+anchor_images_path = cache_dir / "left"
+positive_images_path = cache_dir / "right"
+
+"""shell
+gdown --id 1jvkbTr_giSP3Ru8OwGNCg6B4PvVbcO34
+gdown --id 1EzBZUb_mh_Dp_FKD0P4XiYYSd0QBH5zW
+unzip -oq left.zip -d $cache_dir
+unzip -oq right.zip -d $cache_dir
+"""
+
+"""
+## Preparing the data
+
+We are going to use a `tf.data` pipeline to load the data and generate the triplets that we
+need to train the Siamese network.
+
+We'll set up the pipeline using a zipped list with anchor, positive, and negative filenames as
+the source. The pipeline will load and preprocess the corresponding images.
+"""
+
+
+def preprocess_image(filename):
+    """
+    Loads the specified file as a JPEG image, preprocess it and 
+    resizes it to the target shape.
+    """
+
+    image_string = tf.io.read_file(filename)
+    image = tf.image.decode_jpeg(image_string, channels=3)
+    image = tf.image.convert_image_dtype(image, tf.float32)
+    image = tf.image.resize(image, target_shape)
+    return image
+
+
+def preprocess_triplets(anchor, positive, negative):
+    """
+    Given the filenames corresponding to the three images, it loads and
+    preprocess them.
+    """
+
+    return (
+        preprocess_image(anchor),
+        preprocess_image(positive),
+        preprocess_image(negative),
+    )
+
+
+"""
+Let's setup our pipeline using a zipped list with anchor and positive
+filenames as the source. The output of our pipeline contains a triplet with
+the anchor, the positive, and the negative image.
+"""
+
+anchor_dataset = tf.data.Dataset.list_files(
+    file_pattern=str(anchor_images_path / "*.jpg"), shuffle=False
+)
+positive_dataset = tf.data.Dataset.list_files(
+    file_pattern=str(positive_images_path / "*.jpg"), shuffle=False
+)
+
+# The negative sample is a randomly selected image from either the anchor or
+# the positive list of images. To ensure that we select a random image, we need
+# to shuffle the dataset.
+negative_dataset = anchor_dataset.concatenate(positive_dataset)
+negative_dataset = negative_dataset.shuffle(buffer_size=20000)
```

Comment:
For randomness, I recommend:

1. First, globally shuffle the filenames before turning them into a dataset:
```
rng = np.random.RandomState(seed_1)
rng.shuffle(anchor_images_paths)
rng.shuffle(positive_images_paths)

negative_paths = anchor_images_paths + positive_images_paths
np.random.RandomState(seed_2).shuffle(negative_paths)
```

2. Use a mid-sized buffer to shuffle as part of the dataset itself (e.g. 4096 above, 1024 below)